docs: add lambda pre-aggs and ksqldb (#5321)
cube-js/cube.js: docs/content/Caching/Lambda-Pre-Aggregations.mdx

---
title: Lambda Pre-Aggregations
permalink: /caching/pre-aggregations/lambda-pre-aggregations
category: Caching
menuOrder: 4
---

> Lambda pre-aggregations follow the [Lambda architecture](https://en.wikipedia.org/wiki/Lambda_architecture) design to union real-time and batch data. Cube acts as a serving layer and uses pre-aggregations as a batch layer and source data or other pre-aggregations, usually [streaming][streaming-pre-agg], as a speed layer.

Reviewer: Would it be better to say "service layer" (vs. "serving layer")?

Author: It is specific term used in the lambda architecture - https://en.wikipedia.org/wiki/Lambda_architecture

<WarningBox>

Lambda pre-aggregations only work with Cube Store. 

> Additionally, we’re going to remove support for external storages, other than Cube Store, later this year. Cube Store will replace Redis and, therefore will be a required component to run Cube even without pre-aggregations.

Reviewer: ```suggestion
Additionally, we’re going to remove support for external storages, other than Cube Store, later this year. [Cube Store will replace Redis](https://cube.dev/blog/replacing-redis-with-cube-store) and, therefore will be a required component to run Cube even without pre-aggregations.
```

</WarningBox>

## Use cases

Below we are looking at the most common examples of using lambda pre-aggregations.

### Batch and source data

> Batch data is coming from pre-aggregation and real-time data is coming from the data source.

Reviewer 2: ```suggestion
In this scenario, batch data comes from a pre-aggregation and real-time data comes from the data source.
```

<div style="text-align: center">
  <img
    alt="Lambda pre-aggregation batch and source diagram"
    src="https://raw.githubusercontent.com/cube-js/cube.js/master/docs/content/Caching/lambda-batch-source.png"
    style="border: none"
    width="100%"
  />
</div>

First, you need to create pre-aggregations that will contain your batch data. In the following example, we call it **batch.** Please note, it must have `timeDimension`, Cube will use it to union batch data with source data.

You control the batch part of your data with `buildRangeStart` and `buildRangeEnd` properties of pre-aggregation to determine specific window for your batched data.

Next, you need to create a lambda pre-aggregation. To do that, create pre-aggregation with type `rollupLambda`, specify rollups you would like to use with `rollups` property, and finally set `unionWithSourceData: true` to use source data as a real-time layer.

> Please make sure that lambda pre-aggreation definition goes first when defining your pre-aggregations.

Reviewer: ```suggestion
Please make sure that the lambda pre-aggreation definition comes first when defining your pre-aggregations.
```

> Please make sure that the lambda pre-aggreation definition comes first when defining your pre-aggregations.

Reviewer 2: ```suggestion
Please make sure that the lambda pre-aggregation definition comes first when defining your pre-aggregations.

```

```js
lambda: {
  type: `rollupLambda`,
  unionWithSourceData: true,
  rollups: [Users.batch]
},
batch: {
  measures: [Users.count],
  dimensions: [Users.name],
  timeDimension: Users.createdAt,
  granularity: `day`,
  buildRangeStart: {
    sql: `SELECT '2020-01-01'`
  },
  buildRangeEnd: {
    sql: `SELECT '2022-05-30'`
  }
}
```

### Batch and streaming data

> Batch data is coming from one pre-aggregation and real-time data coming from another, [streaming pre-aggregation][streaming-pre-agg].

Reviewer 2: ```suggestion
In this scenario, batch data is comes from one pre-aggregation and real-time data comes from a [streaming pre-aggregation][streaming-pre-agg].
```

<div style="text-align: center">
  <img
    alt="Lambda pre-aggregation batch and streaming diagram"
    src="https://raw.githubusercontent.com/cube-js/cube.js/master/docs/content/Caching/lambda-batch-streaming.png"
    style="border: none"
    width="100%"
  />
</div>

You can use lambda pre-aggregations to combine data from multiple pre-aggregation, where one pre-aggregation can have batch data and another streaming.

```js

> batchStreamingLambda: {

Reviewer 2: Could we add the `cube()` definition for this example? It looks like the example uses two cubes, but its not clear which one the pre-aggregation definition lives in

Author: Yes, let's do it! It is a good idea. Can you please send it as a separate PR?
